homeassistant/components/light/mysensors.py

```diff
@@ -109,118 +100,258 @@ def device_state_attributes(self):
         return device_attr
 
     @property
+    def available(self):
+        """Return True if entity is available."""
+        return self.value_type in self._values
+
+    @property
     def is_on(self):
         """True if device is on."""
         return self._state
 
     def turn_on(self, **kwargs):
         """Turn the device on."""
+        if self.gateway.optimistic:
+            # optimistically assume that light has changed state
+            self.update_ha_state()
+
+    def turn_off(self, **kwargs):
+        """Turn the device off."""
+        value_type = kwargs.get('value_type')
+        value = kwargs.get('value')
+        if value_type is not None and value is not None:
+            self.gateway.set_child_value(
+                self.node_id, self.child_id, value_type, value)
+        else:
+            _LOGGER.warning(
+                '%s: value_type %s, value = %s, '
+                'None is not valid argument when setting child value'
+                '', self._name, value_type, value)
+        if self.gateway.optimistic:
+            # optimistically assume that light has changed state
+            self._state = False
+            self.update_ha_state()
+
+    def update(self):
+        """Update the controller with the latest value from a sensor."""
+        node = self.gateway.sensors[self.node_id]
+        child = node.children[self.child_id]
+        self.battery_level = node.battery_level
+        for value_type, value in child.values.items():
+            _LOGGER.debug(
+                '%s: value_type %s, value = %s', self._name, value_type, value)
+            self._values[value_type] = value
+
+
+class MySensorsLightLight(MySensorsLight):
+    """Light child class to MySensorsLight."""
+
+    def __init__(self, *args):
+        """Setup instance attributes."""
+        super().__init__(*args)
+
+    def turn_on(self, **kwargs):
+        """Turn the device on."""
         set_req = self.gateway.const.SetReq
-        rgb = self._rgb
-        brightness = self._brightness
-        white = self._white
 
-        if set_req.V_LIGHT in self._values and not self._state:
+        if not self._state:
             self.gateway.set_child_value(
                 self.node_id, self.child_id, set_req.V_LIGHT, 1)
 
-        if ATTR_BRIGHTNESS in kwargs and set_req.V_DIMMER in self._values and \
+        if self.gateway.optimistic:
+            # optimistically assume that light has changed state
+            self._state = True
+        super().turn_on(**kwargs)
```

Comment:
I don't like how super is being used here. Why not extract a method called `optimistic_update` that will only update if necessary (ie `gateway.optimistic == True`) and call this from `turn_on` and `turn_off`